<html>
    <head>
        <title>Async/Await test page</title>
        <meta name="description" content="Test how your browser handles the async and await keywords defined in ES2017." />
        <link rel="stylesheet" type="text/css" href="css/common.css">
    </head>
    <body>
        <p><a href=".." title="Test pages">&lt;&lt; Test pages</a></p>
        <h1>Async/Await test page</h1>
        <p>
            This page tests how your browser handles the async and await
            keywords defined in ES2017. A function can be declared async. This
            means the function returns a Promise which may resolve later. Within
            an async function the await keyword can be used. The await keyword
            pauses the execution of the async function until the value awaited
            on resolves, or rejects. A resolved value is returned by the await
            keyword and a reject value is thrown. Together these keywords allow
            asynchronous code to be written as though it was synchronous.
        </p>
        <h2>Results</h2>
        <noscript><p>JavaScript is not supported or disabled in your browser.
            There is no chance of this working.</p></noscript>
        <div id="script-error"></div>
        <div id="results" style="display: none;">
            <p>
                <table cellspacing="0">
                    <tr><th colspan="3">This table shows the result of awaiting on
                        different values in an async method</th></tr>
                    <tr>
                        <th>Value awaited on</th>
                        <th>The value returned by await</th>
                        <th>The value thrown by await</th>
                    <tr>
                    <tr>
                        <td>Promise resolved to 5</td>
                        <td id="result-0"></td>
                        <td id="error-0"></td>
                    </tr>
                    <tr>
                        <td>Promise rejected with an error</td>
                        <td id="result-1"></td>
                        <td id="error-1"></td>
                    </tr>
                    <tr>
                        <td>Literal 5</td>
                        <td id="result-2"></td>
                        <td id="error-2"></td>
                    </tr>
                    <tr>
                        <td>An error</td>
                        <td id="result-3"></td>
                        <td id="error-3"></td>
                    </tr>
                    <tr>
                        <td>Promise-like resolved to 5</td>
                        <td id="result-4"></td>
                        <td id="error-4"></td>
                    </tr>
                    <tr>
                        <td>Promise-like rejected with an error</td>
                        <td id="result-5"></td>
                        <td id="error-5"></td>
                    </tr>
                </table>
            </p>
            <p>
                It is worth noting that Errors awaited on are not thrown they are
                returned as successful results of await.
            </p>
            <p>
                A Promise-like object is an object with a then method but that is
                not a promise.
            </p>
        </div>

        <script>
            // Set text indicating that async is not supported
            document.getElementById('script-error').innerHTML =
                '<p>The script could not be parsed, async/await not ' +
                'supported.</p>';
        </script>

        <script>
            // Clear the text indicating that async is not supported
            document.getElementById('script-error').innerHTML = '';
            document.getElementById('results').style.display = 'block';

            // The values used by the test
            var value = 5;
            var resolvedPromise = new Promise(function(resolve, reject) {
                resolve(value);
            });
            var resolvedPromiseLike = {
                then: function(resolve, reject) {
                    resolve(value);
                }
            };
            var error = new Error('An error');
            var rejectedPromise = new Promise(function(resolve, reject) {
                reject(error);
            });
            var rejectedPromiseLike = {
                then: function(resolve, reject) {
                    reject(error);
                }
            };

            // The asynchronous function using await
            async function awaitOn(testId, value) {
                try {
                    var rId = 'result-' + testId;
                    document.getElementById(rId).innerText = await value;
                }
                catch (e) {
                    var eId = 'error-' + testId;
                    document.getElementById(eId).innerText = e;
                }
            }

            // Pass different values to the asynchronous function to await on
            awaitOn(0, resolvedPromise);
            awaitOn(1, rejectedPromise);
            awaitOn(2, value);
            awaitOn(3, error);
            awaitOn(4, resolvedPromiseLike);
            awaitOn(5, rejectedPromiseLike);
        </script>
    </body>
</html>
